<h1>Nielson - Senior Data Scientist test assignment</h1>

<p>This notebook will be used for the <code>test assignment</code> provided by <strong>Nielson</strong> as part of interview process for the <strong>Senior Data Scientist position.</strong></p>

<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li><a href="imp">Import libraries and install packages</a></li>
        <li><a href="load">Load and Understand Data</a></li>
        <li><a href="wrangle">Data and Features Preparation</a></li>
        <li><a href="model">Data Modelling</a></li>
        <li><a href="vis">Visualization</a></li>
    </ul>
    
</div>

<hr>

<h2 id="imp">Import Libraries</h2>
<p>Get all required libraries and packages for Python</p>

In [2]:
import pandas as pd
import numpy as np
from scipy import stats
import matplotlib as plt
import matplotlib.pylab as plt
from matplotlib import pyplot
import sys

import matplotlib.pyplot as plt

from sklearn import linear_model
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics

%matplotlib inline
print('Libraries installed and imported!')

Libraries installed and imported!


<hr>
<h2 id="load">Data Load and Understanding</h2>
<p> Load data from <strong>CSV</strong> file and do a preliminary understanding/cleansing.</p>

<h3>Load data from 2 CSV sources</h3>

In [4]:
#Read CSV file into Panda dataframe
read_path = "dt_challenge.csv"
#List of russian encoding traversed: cp1251, iso8859_5, koi8_r, mac_cyrillic, cp866, cp855 
#from https://docs.python.org/3/library/codecs.html#standard-encodings
df = pd.read_csv(read_path, encoding='cp1251')
df.head(10)

Store_id           Region        barcode  TransactionID        Date  \
0         1  Санкт-Петербург  3083680538404     1138789911  2018-08-22   
1         1  Санкт-Петербург       40084107     1214906739  2018-09-03   
2         1  Санкт-Петербург       40084107     1214906739  2018-09-03   
3         1  Санкт-Петербург       40084107     1214906739  2018-09-03   
4         1  Санкт-Петербург       40084107     1322925912  2018-09-22   
5         1  Санкт-Петербург       40084107     1322925912  2018-09-22   
6         1  Санкт-Петербург       40084107     1322925912  2018-09-22   
7         1  Санкт-Петербург       40084107     1236599978  2018-09-07   
8         1  Санкт-Петербург       40084107     1236599978  2018-09-07   
9         1  Санкт-Петербург       40084107     1236599978  2018-09-07   

                                             barname  sales_items  sales_value  
0          ГОРОШЕК """"Bonduelle"""" НЕЖНЫЙ 200Г Ж/Б          1.0         86.0  
1  Д Kinder Chocolate ЯЙЦО KINDER СЮРПРИЗ КИНДЕР ...          1.0         90.0  
2  Д Kinder Chocolate ЯЙЦО KINDER СЮРПРИЗ КИНДЕР ...          1.0         90.0  
3  Д Kinder Chocolate ЯЙЦО KINDER СЮРПРИЗ КИНДЕР ...          1.0         90.0  
4  Д Kinder Chocolate ЯЙЦО KINDER СЮРПРИЗ КИНДЕР ...          2.0        180.0  
5  Д Kinder Chocolate ЯЙЦО KINDER СЮРПРИЗ КИНДЕР ...          2.0        180.0  
6  Д Kinder Chocolate ЯЙЦО KINDER СЮРПРИЗ КИНДЕР ...          2.0        180.0  
7  Д Kinder Chocolate ЯЙЦО KINDER СЮРПРИЗ КИНДЕР ...          1.0         90.0  
8  Д Kinder Chocolate ЯЙЦО KINDER СЮРПРИЗ КИНДЕР ...          1.0         90.0  
9  Д Kinder Chocolate ЯЙЦО KINDER СЮРПРИЗ КИНДЕР ...          1.0         90.0

In [5]:
df.tail(10)

Store_id               Region        barcode  TransactionID  \
7970607       518  Ставропольский край  8996001414002     1036162714   
7970608       518  Ставропольский край  8996001414002     1104556699   
7970609       518  Ставропольский край  8996001414002     1317060928   
7970610       518  Ставропольский край  8996001414002     1086254409   
7970611       518  Ставропольский край  8996001414002     1312395097   
7970612       518  Ставропольский край  9000100321648     1331475044   
7970613       518  Ставропольский край  9000100561501     1256393365   
7970614       518  Ставропольский край  9000100561501     1256393365   
7970615       518  Ставропольский край  9000101029260     1260902572   
7970616       518  Ставропольский край       96032961      956798407   

               Date                                            barname  \
7970607  2018-08-04         Капучино. Торабика 25 г.с шоколад. крошкой   
7970608  2018-08-16         Капучино. Торабика 25 г.с шоколад. крошкой   
7970609  2018-09-21         Капучино. Торабика 25 г.с шоколад. крошкой   
7970610  2018-08-12         Капучино. Торабика 25 г.с шоколад. крошкой   
7970611  2018-09-21         Капучино. Торабика 25 г.с шоколад. крошкой   
7970612  2018-09-24                              Порошок  """"Лоск""""   
7970613  2018-09-11                              Порошок  """"Дени""""   
7970614  2018-09-11                              Порошок  """"Дени""""   
7970615  2018-09-12  Кондеционер""""  Вернел""""   910мл Арома Орхи...   
7970616  2018-07-21                                         Дезидорант   

         sales_items  sales_value  
7970607          3.0         35.0  
7970608          1.0         12.0  
7970609          1.0         17.0  
7970610          3.0         35.0  
7970611          1.0         17.0  
7970612          1.0         93.0  
7970613          1.0         68.0  
7970614          1.0         68.0  
7970615          1.0        171.0  
7970616          1.0        135.0

In [12]:
#Clean data: barname, strip spaces and strange characters
df['barname'] = df['barname'].str.strip()
df['barname'] = df['barname'].str.replace('\W', '')
df['barname'].value_counts()

#!Alert!: maybe not such a good idea at this stage. Consider removing to feature-engineering section where DF's are transposed 

RothmansDemiМРЦ1000010000                                         17255
БОНДКомпактБлю                                                    16633
КофеНескафе3в1васс85016г1000                                      15883
жрОрбитасс3000                                                    15292
сигБондпремиуммикс                                                14786
                                                                  ...  
МультифруктнапитоксилгазСтавроп15л                                    1
САЛФЕТКАУНИВЕРСАЛЬНЫЕФРЕКЕНБОК2ШТ                                     1
ЗаправкаМАГГИдлящей250гр                                              1
СырокглазированныйСибирячок38гвассортименте                           1
ПриправычСмесьпряностейдляпикантныхкуриныхкрылышекспакетом30гр        1
Name: barname, Length: 250966, dtype: int64

In [7]:
df.shape

(7970617, 8)

In [8]:
#Read barcode with categories CSV file into Panda dataframe
#List of russian codes and categories downloaded from https://github.com/papyrussolution/UhttBarcodeReference/releases

read_path = "uhtt_barcode_ref_all.csv"

ean = pd.read_csv(read_path, encoding='utf8', engine='c', sep='\t')
ean.head()

/Applications/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (0,6) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


ID         UPCEAN                                               Name  \
0  1850278    61500127178          !ajua! caffeine free soda Mandarin Orange   
1  1850279    61500127161             !ajua! caffeine free soda PINEAPPLE 1L   
2  1850277    61500127154  !ajua! caffeine free soda tutti fruitti Fruit ...   
3  3604539  4603726031011            !DEAS APPL&CAR&BEET DIET 100% V 1L BO J   
4  3604540  4603726031004                !DEAS APPL&CELR DIET 100% V 1L BO J   

   CategoryID                                       CategoryName    BrandID  \
0         1.0                       Неклассифицированные/default        0.0   
1         1.0                       Неклассифицированные/default        0.0   
2         1.0                       Неклассифицированные/default        0.0   
3     17374.0  Продукты питания (folder)/Напитки безалкогольн...  3604538.0   
4     17374.0  Продукты питания (folder)/Напитки безалкогольн...  3604538.0   

  BrandName  
0       NaN  
1       NaN  
2       NaN  
3     !DEAS  
4     !DEAS

In [9]:
ean['Name'].value_counts()

Nieuwkoop dual top / chameleon                                                                    4675228
Estee Lauder eau private collection 1.7 EDP l                                                       86287
Jonnesway ae300025 приспособление для отбора технических жидкостей. пневматический привод. 16л      80271
Неклассифицированные/default                                                                           44
Инструменты (folder)/Инструмент                                                                        10
                                                                                                   ...   
Подгузники insinse v6 трусики (13-15кг) 18шт супертонкие xl (Золотые) 12шт                              1
Sephora beauty+ Крем увлажняющий для лица 30мл                                                          1
Carlton coasters #2                                                                                     1
Manhattan блеск-сияние для губ т.461 high Shin

In [10]:
ean.shape    

(8675083, 7)

In [11]:
#Clean data: Name, strip spaces and strange characters
ean['Name'] = ean['Name'].str.strip()
ean['Name'] = ean['Name'].str.replace('\W', '')
ean['Name'].value_counts()

#!Alert!: maybe not such a good idea at this stage. Consider removing to feature-engineering section where DF's are transposed 

Nieuwkoopdualtopchameleon                                                              4675228
EsteeLaudereauprivatecollection17EDPl                                                    86287
Jonneswayae300025приспособлениедляотборатехническихжидкостейпневматическийпривод16л      80271
Неклассифицированныеdefault                                                                 44
ИнструментыfolderИнструмент                                                                 10
                                                                                        ...   
Angiespopcornboomchkaswtslty                                                                 1
Лепешкачиабаттазерновая                                                                      1
Panasonicbatterylithum                                                                       1
Подсвечникизстекла                                                                           1
Фисташкиореховыймиржареныесоленые70гму            

In [19]:
ean['CategoryName'].value_counts()

Неклассифицированные/default                                                                                                          1237866
Продукты питания (folder)/Продукты питания                                                                                             193445
Продукты питания (folder)/Алкогольные напитки/Вино (folder)/Вино                                                                        53018
Косметика (folder)/Косметика                                                                                                            41289
Лекарственные средства (folder)/Лекарственные средства                                                                                  40202
                                                                                                                                       ...   
Держатели телефона, GPS, коммуникатора, AUTOVIRAZH/Держатели сотового телефона, коммуникатора, GPS//                                        1
Продук

In [20]:
ean['BrandName'].value_counts()

Gloria Jeans        65702
Cooper Bussmann     18987
Tdm                 16617
Bradford White      14084
Salewa              11173
                    ...  
Deaths Door             1
ВЕЧНЫЙ ЕРЕВАН           1
Huggies Pull-ups        1
Purdys                  1
Swissflame              1
Name: BrandName, Length: 67141, dtype: int64

<hr>
<h3>Left outer join transaction records with data on EAN categories, drop unused colums</h3>

In [15]:
ean = ean.sort_values(['UPCEAN', 'CategoryID', 'BrandID'], ascending=[True, True, True])

In [16]:
ean.tail(20)

ID         UPCEAN  \
6335477      2187678  9999991758278   
6335476      2187677  9999991758285   
7383273  3.34981e+06  9999996002635   
7223205      3321454  9999997299805   
7383277  3.34981e+06  9999997560295   
8048372      2914746  9999999000232   
8258677        21252  9999999019234   
8258678        21253  9999999019241   
8258958        21248  9999999019524   
8258679        21255  9999999067976   
38421        2016036  9999999170454   
8173396       113631  9999999214431   
7383248  3.34979e+06  9999999229060   
7918221  3.45796e+06  9999999229763   
7383438  3.34991e+06  9999999231872   
8173398      1697694  9999999232817   
7329186        63984  9999999254499   
8173399      1697695  9999999292736   
7556648       746487  9999999309779   
7469883  3.36805e+06  9999999519277   

                                                     Name  CategoryID  \
6335477                            Prensadograntoro6x5420         1.0   
6335476                             Prensadodoublet6x6220         1.0   
7383273                       Коверimperialбербер80x150см         1.0   
7223205                      Зонтпляжный17мснаклоном37779    285706.0   
7383277  Коверomega250x350овал4152gdrb24светлосерыйчерный         1.0   
8048372        ПомадагубнаяGuerlainkisskiss301карамель35г      1069.0   
8258677                    Скатертьпвхсокант110x140см6658    523282.0   
8258678                    Скатертьпвхсокант130x160см6659    523282.0   
8258958                     Скатертьхбсокант130x160см6663    523282.0   
8258679                    Скатертьпвхсокант140x220см6669    523282.0   
38421                                    Afselpriv1mad25s         1.0   
8173396                                  Салфетницаск5072     38328.0   
7383248                                        Ковер15x23         1.0   
7918221                                        Палас15x23         1.0   
7383438                                       Коврик08x15    315162.0   
8173398                             Салфетницаск526853162     38328.0   
7329186                       Кастрюляполевыецветы3л16124    314257.0   
8173399                             Салфетницаск526853163     38328.0   
7556648                             Крышкарегентd28смri28   2420630.0   
7469883                       КоньякHennessyвсоп40005лстб     22210.0   

                                              CategoryName   BrandID BrandName  
6335477                       Неклассифицированные/default       0.0       NaN  
6335476                       Неклассифицированные/default       0.0       NaN  
7383273                       Неклассифицированные/default       0.0       NaN  
7223205                           Галантерея (folder)/Зонт       0.0       NaN  
7383277                       Неклассифицированные/default       0.0       NaN  
8048372                   Косметика (folder)/Помада губная       0.0       NaN  
8258677             Хозяйственные товары (folder)/Скатерть   21246.0    ВИЛИНА  
8258678             Хозяйственные товары (folder)/Скатерть   21246.0    ВИЛИНА  
8258958             Хозяйственные товары (folder)/Скатерть   21246.0    ВИЛИНА  
8258679             Хозяйственные товары (folder)/Скатерть   21246.0    ВИЛИНА  
38421                         Неклассифицированные/default       0.0       NaN  
8173396  Посуда (folder)/Посуда столовая, предметы серв...       0.0       NaN  
7383248                       Неклассифицированные/default       0.0       NaN  
7918221                       Неклассифицированные/default       0.0       NaN  
7383438               Хозяйственные товары (folder)/Коврик       0.0       NaN  
8173398  Посуда (folder)/Посуда столовая, предметы серв...       0.0       NaN  
7329186                           Посуда (folder)/Кастрюля       0.0       NaN  
8173399  Посуда (folder)/Посуда столовая, предметы серв...       0.0       NaN  
7556648                    Посуда (folder)/Посуда кухонная  122990.0    Регент  
7469883  Продукты питания (folder)/Алкогольн

In [17]:
#sort ean by UPCEAN, remove duplicates (to ensure left_merged dataframe has same # raws as df)
ean = ean.drop_duplicates(subset='UPCEAN', keep="last")
ean.head(20)

ID  UPCEAN  \
6160384   structure ral: outside-diam=60 height=90 wt=15       1   
8431371                                           548072      17   
7967124                                           548330      24   
7208388                                           549326      31   
8345972                                           549356      48   
8308600                                           548385      55   
8345896                                           330919      86   
8345328                                           548416      93   
7439478                                           549416     109   
7675680                                           547654     116   
7208389                                           549327     123   
8641012                                           549160     130   
8674192                                           548353     154   
6893358                                              NaN     163   
7758604                                           549322     178   
6823258                                           549417     192   
8673773                                           549478     277   
8673774                                           549456     284   
8673775                                           549479     291   
7463055                                           549432     307   

                                                      Name  CategoryID  \
6160384                        Неклассифицированныеdefault   3712434.0   
8431371                  ТортшоколадныйБриз250гКремварсгущ    476443.0   
7967124                Печеньебккмосксдобнфантазиявишня5кг     21865.0   
7208388                 Зажигалкапьезоluxexhd101wpsurfing2    285556.0   
8345972  СухарикигренкиЕмеляассвмасле8миксчеснперсолтом...   1202446.0   
8308600                    Спокойнойночидражевпляйце5г6x24         1.0   
8345896          Сухарикивоздсолбомбастербарбекю40г30бс332   1202446.0   
8345328                 Сухарихлебныйдом4502изюминка1с350г    676644.0   
7439478  Консервырыбныеморскоесодружествокильканераздвт...        48.0   
7675680              Маслоподсолнечноедарыкубанинераф065л9    184383.0   
7208389              Зажигалкапьезоluxexhd101swpstrawberry    285556.0   
8641012  Шоколадноеяйцомолочншоколадмеладасигрушкойcraz...         1.0   
8674192                     Ятоматыдввтоматномсоусе680гсб8         1.0   
6893358  ХозяйственныетоварыfolderБытоваяхимияfolderСре...     18446.0   
7758604                        Нккардвойныесливкималина4кг         1.0   
6823258                WinxClubкарвкольцеGoodфрукт29г12x12         1.0   
8673773             Якпфконсенсуспеченьеглаголикикласс250г     21865.0   
8673774              Якпфконсенсуспеченьекокосанкаорех200г     21865.0   
8673775                    Якпфконсенсуспеченьекурабье250г     21865.0   
7463055              Конфетыневскийкондитермияйогуртнач3кг     22145.0   

                                              CategoryName  BrandID  \
6160384                                          Nieuwkoop      NaN   
8431371  Продукты питания (folder)/Кондитерские изделия...      0.0   
7967124  Продукты питания (folder)/Кондитерские изделия...      0.0   
7208388            Хозяйственные товары (folder)/Зажигалка      0.0   
8345972                 Продукты питания (folder)/Сухарики      0.0   
8308600                       Неклассифицированные/default      0.0   
8345896                 Продукты питания (folder)/Сухарики      0.0   
8345328   Продукты питания (folder)/Сухари, сушки, баранки  37625.0   
7439478  Продукты питания (folder)/Рыба и морепродукты ...      0.0   
7675680  Продукты питания (folder)/Масложировая продукц...      0.0   
7208389            Хозяйственные товары (folder)/Зажигалка      0.0   
8641012                       Неклассифицированные/default      0.0   
8674192                       Неклассифицированные/default      0.0   
6893358                                           NORDLAND      NaN   
77586

In [18]:
merged_left_out = pd.merge(left=df,right=ean, how='left', left_on='barcode', right_on='UPCEAN')
merged_left_out.shape

(7970617, 15)

In [21]:
merged_left_out.head()

Store_id           Region        barcode  TransactionID        Date  \
0         1  Санкт-Петербург  3083680538404     1138789911  2018-08-22   
1         1  Санкт-Петербург       40084107     1214906739  2018-09-03   
2         1  Санкт-Петербург       40084107     1214906739  2018-09-03   
3         1  Санкт-Петербург       40084107     1214906739  2018-09-03   
4         1  Санкт-Петербург       40084107     1322925912  2018-09-22   

                                             barname  sales_items  \
0                       ГОРОШЕКBonduelleНЕЖНЫЙ200ГЖБ          1.0   
1  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
2  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
3  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
4  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          2.0   

   sales_value     ID        UPCEAN  \
0         86.0  22661  3.083681e+12   
1         90.0  30127  4.008411e+07   
2         90.0  30127  4.008411e+07   
3         90.0  30127  4.008411e+07   
4        180.0  30127  4.008411e+07   

                                                Name  CategoryID  \
0              ГорошекзеленыйBonduelle200гх12Венгрия        65.0   
1  Шоколадноеяйцоferrerokinderсюрприз20гт1ассорти...     19677.0   
2  Шоколадноеяйцоferrerokinderсюрприз20гт1ассорти...     19677.0   
3  Шоколадноеяйцоferrerokinderсюрприз20гт1ассорти...     19677.0   
4  Шоколадноеяйцоferrerokinderсюрприз20гт1ассорти...     19677.0   

                                        CategoryName  BrandID       BrandName  
0  Продукты питания (folder)/Консервы (folder)/Ко...  22653.0       Bonduelle  
1  Продукты питания (folder)/Кондитерские изделия...  30123.0  Kinder Сюрприз  
2  Продукты питания (folder)/Кондитерские изделия...  30123.0  Kinder Сюрприз  
3  Продукты питания (folder)/Кондитерские изделия...  30123.0  Kinder Сюрприз  
4  Продукты питания (folder)/Кондитерские изделия...  30123.0  Kinder Сюрприз

In [30]:
#Removing the credit transactions in this analysis. For future improvement: 
# or remove complete credit/debit pair for more accurate results 
# keep both and analyse separately for churn trends/insights. 

merged_left_out = merged_left_out[(merged_left_out['sales_items']>0)& (merged_left_out['sales_value']>0)]
merged_left_out.shape

(7966550, 12)

In [22]:
merged_left_out.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7970617 entries, 0 to 7970616
Data columns (total 15 columns):
Store_id         int64
Region           object
barcode          int64
TransactionID    int64
Date             object
barname          object
sales_items      float64
sales_value      float64
ID               object
UPCEAN           float64
Name             object
CategoryID       float64
CategoryName     object
BrandID          float64
BrandName        object
dtypes: float64(5), int64(3), object(7)
memory usage: 973.0+ MB


In [23]:
merged_left_out['barname'].value_counts()

RothmansDemiМРЦ1000010000                                         17255
БОНДКомпактБлю                                                    16633
КофеНескафе3в1васс85016г1000                                      15883
жрОрбитасс3000                                                    15292
сигБондпремиуммикс                                                14786
                                                                  ...  
МультифруктнапитоксилгазСтавроп15л                                    1
САЛФЕТКАУНИВЕРСАЛЬНЫЕФРЕКЕНБОК2ШТ                                     1
ЗаправкаМАГГИдлящей250гр                                              1
СырокглазированныйСибирячок38гвассортименте                           1
ПриправычСмесьпряностейдляпикантныхкуриныхкрылышекспакетом30гр        1
Name: barname, Length: 250966, dtype: int64

In [24]:
merged_left_out['BrandName'].value_counts()

Winston                               429692
Rothmans                              312390
Philip Morris                         269880
Bond Street                           246931
JTI L-D                               172205
                                       ...  
CHASAR                                     1
BELOSNEZHNAYA (SITNO/MAGNITOGORSK)         1
Maxell                                     1
Эдванс                                     1
PACO RABANNE                               1
Name: BrandName, Length: 8658, dtype: int64

In [25]:
merged_left_out['CategoryName'].value_counts()

Табакокурение (folder)/Сигареты                                                                3279407
Продукты питания (folder)/Алкогольные напитки/Пиво                                              587149
Неклассифицированные/default                                                                    365651
Продукты питания (folder)/Напитки безалкогольные/Напитки                                        346642
Продукты питания (folder)/Мороженое                                                             298948
                                                                                                ...   
Галантерея (folder)/Косметические аксессуары                                                         1
Продукты питания (folder)/Алкогольные напитки/Вино (folder)/Вина Франции игристые и Шампань          1
Гигиена (folder)/Уход за зубами и ртом (folder)/Крем для зубных протезов фиксирующий                 1
средства по уходу и очищению кожи ребёнка                                

In [26]:
merged_left_out['Region'].value_counts()

Воронежская область                998616
Московская область                 531620
Москва                             374287
Томская область                    344739
Краснодарский край                 312306
                                    ...  
Еврейская автономная область         8177
Республика Алтай                     6494
Тамбовская область                   3202
Кабардино-Балкарская Республика      3097
Ямало-Ненецкий автономный округ      2566
Name: Region, Length: 73, dtype: int64

In [27]:
merged_left_out['Store_id'].value_counts()

435    282657
57     245751
144    150307
490     97642
177     88997
        ...  
129       602
405       583
396       499
338       455
305       251
Name: Store_id, Length: 518, dtype: int64

In [28]:
#Print number of null values 
print("Count null's per variable\n", merged_left_out.isnull().sum())

#drop unnecessary colums: ID, UPCEAN, Name
merged_left_out.drop(['ID','UPCEAN','Name'], axis=1, inplace=True)

Count null's per variable
 Store_id              0
Region                0
barcode               0
TransactionID         0
Date                  0
barname            3266
sales_items           0
sales_value           2
ID               200733
UPCEAN           200733
Name             200733
CategoryID       200733
CategoryName     200733
BrandID          200733
BrandName        898825
dtype: int64


In [31]:
merged_left_out.shape

(7966550, 12)

<div class="alert alert-block alert-success">
<b>Data Load Complete</b> 
    <ul> 
        <li>Load assignment CSV, pre-process: <code>df [7970617, 8]</code></li>
        <li>Load additional EAN to Category reference CSV, pre-process: <code>ean [8675083, 7]</code></li>
        <li>Left join two sources, clean, drop unused, count categorical occurences: <br>
            <code>merged_left_out [7966550, 12]</code> <br>
            <code>Region: 73 </code> <br>
            <code>Store_id: 518 </code> <br>
            <code>CategoryName: 516 </code> <br>
            <code>BrandName: 8658 </code> <br>
            <code>barname: 250966 </code> <br>
        </li>
    </ul>
</div>

<h3>Clean data (I): Fill N/A cells, standardize numerical values, drop N/A records</h3>

In [32]:
#Convert variables (CategoryID, BrandID) from float to integer
merged_left_out["CategoryID"].fillna("0", inplace = True)
merged_left_out["BrandID"].fillna("0", inplace = True) 
merged_left_out["CategoryID"] = merged_left_out["CategoryID"].astype("int")
merged_left_out["BrandID"] = merged_left_out["BrandID"].astype("int")

#Fill the "unknown" for the N/A in the BrandID, BrandName, CategoryID and CategoryName, barname
merged_left_out["barname"].fillna("Unknown Name", inplace = True) 
merged_left_out["CategoryName"].fillna("Unknown Category", inplace = True) 
merged_left_out["BrandName"].fillna("Unknown Brand", inplace = True) 

print("Count null's per variable\n", merged_left_out.isnull().sum())

Count null's per variable
 Store_id         0
Region           0
barcode          0
TransactionID    0
Date             0
barname          0
sales_items      0
sales_value      0
CategoryID       0
CategoryName     0
BrandID          0
BrandName        0
dtype: int64


In [33]:
merged_left_out[(merged_left_out['sales_value'].isnull()==True)]

Empty DataFrame
Columns: [Store_id, Region, barcode, TransactionID, Date, barname, sales_items, sales_value, CategoryID, CategoryName, BrandID, BrandName]
Index: []

In [34]:
merged_left_out.dropna(subset=['sales_value'], axis=0, inplace=True)

print("Count null's per variable\n", merged_left_out.isnull().sum())

Count null's per variable
 Store_id         0
Region           0
barcode          0
TransactionID    0
Date             0
barname          0
sales_items      0
sales_value      0
CategoryID       0
CategoryName     0
BrandID          0
BrandName        0
dtype: int64


<h3>Clean data (II): Split Categories Columns, Impute values for Unknown Barname and Unknown Category Name</h3>

<h4>Split Categories</h4>

In [35]:
# split the CategoryName into multiple columns 

#1. Get the column names and number of columns for the split
max_cat = merged_left_out['CategoryName'].str.count('/').max()
cat_col = []


for i in range(max_cat+1):
    cat_col.append("Category " + str(i+1))
    
cat_col

['Category 1', 'Category 2', 'Category 3', 'Category 4', 'Category 5']

In [36]:
#2. Split by the delimeter "/"
df_split = merged_left_out['CategoryName'].str.split('/', -1, expand=True)
df_split.head()

0                              1                 2  \
0  Продукты питания (folder)              Консервы (folder)  Консервы овощные   
1  Продукты питания (folder)  Кондитерские изделия (folder)           Шоколад   
2  Продукты питания (folder)  Кондитерские изделия (folder)           Шоколад   
3  Продукты питания (folder)  Кондитерские изделия (folder)           Шоколад   
4  Продукты питания (folder)  Кондитерские изделия (folder)           Шоколад   

      3     4  
0  None  None  
1  None  None  
2  None  None  
3  None  None  
4  None  None

In [38]:
 # Add column with Split Categories to original left_merged dataset
df_split.columns = cat_col 
df_split.columns.values
merged_left_out[cat_col] = df_split
merged_left_out.tail(40)

Store_id               Region        barcode  TransactionID  \
7970577       518  Ставропольский край  8887290101004     1232424802   
7970578       518  Ставропольский край  8887290101004     1232424802   
7970579       518  Ставропольский край  8887290101004     1232424802   
7970580       518  Ставропольский край  8887290101004     1194924094   
7970581       518  Ставропольский край  8887290101004     1194924094   
7970582       518  Ставропольский край  8887290101004     1194924094   
7970583       518  Ставропольский край  8887290101004     1194924094   
7970584       518  Ставропольский край  8887290101004     1315185326   
7970585       518  Ставропольский край  8887290101004     1315185326   
7970586       518  Ставропольский край  8887290101004     1315185326   
7970587       518  Ставропольский край  8887290101004     1315185326   
7970588       518  Ставропольский край  8887290101004     1298272649   
7970589       518  Ставропольский край  8887290101004     1298272649   
7970590       518  Ставропольский край  8887290101004     1298272649   
7970591       518  Ставропольский край  8887290101004     1298272649   
7970592       518  Ставропольский край  8887856932943     1218153801   
7970593       518  Ставропольский край  8887856932943     1083836389   
7970594       518  Ставропольский край  8996001414002     1197658386   
7970595       518  Ставропольский край  8996001414002      983434487   
7970596       518  Ставропольский край  8996001414002     1043058898   
7970597       518  Ставропольский край  8996001414002     1138185299   
7970598       518  Ставропольский край  8996001414002     1237969430   
7970599       518  Ставропольский край  8996001414002     1184297980   
7970600       518  Ставропольский край  8996001414002     1091372079   
7970601       518  Ставропольский край  8996001414002     1349043862   
7970602       518  Ставропольский край  8996001414002      990611502   
7970603       518  Ставропольский край  8996001414002     1350301148   
7970604       518  Ставропольский край  8996001414002     1241796933   
7970605       518  Ставропольский край  8996001414002     1075997149   
7970606       518  Ставропольский край  8996001414002     1088873137   
7970607       518  Ставропольский край  8996001414002     1036162714   
7970608       518  Ставропольский край  8996001414002     1104556699   
7970609       518  Ставропольский край  8996001414002     1317060928   
7970610       518  Ставропольский край  8996001414002     1086254409   
7970611       518  Ставропольский край  8996001414002     1312395097   
7970612       518  Ставропольский край  9000100321648     1331475044   
7970613       518  Ставропольский край  9000100561501     1256393365   
7970614       518  Ставропольский край  9000100561501     1256393365   
7970615       518  Ставропольский край  9000101029260     1260902572   
7970616       518  Ставропольский край       96032961      956798407   

               Date                                        barname  \
7970577  2018-09-07                                        Кофе3в1   
7970578  2018-09-07                                        Кофе3в1   
7970579  2018-09-07                                        Кофе3в1   
7970580  2018-08-31                                        Кофе3в1   
7970581  2018-08-31                                        Кофе3в1   
7970582  2018-08-31                                        Кофе3в1   
7970583  2018-08-31                                        Кофе3в1   
7970584  2018-09-21                                        Кофе3в1   
7970585  2018-09-21                                        Кофе3в1   
7970586  2018-09-21                                        Кофе3в1   
7970587  2018-09-21                                        Кофе3в1   
7970588  2018-09-18                                        Кофе3в1   
7970589  2018-09-18                                        Кофе3в1   
7970590  2018-09-18                                        Кофе3в1   
797059

In [39]:
merged_left_out.head(40)

Store_id           Region        barcode  TransactionID        Date  \
0          1  Санкт-Петербург  3083680538404     1138789911  2018-08-22   
1          1  Санкт-Петербург       40084107     1214906739  2018-09-03   
2          1  Санкт-Петербург       40084107     1214906739  2018-09-03   
3          1  Санкт-Петербург       40084107     1214906739  2018-09-03   
4          1  Санкт-Петербург       40084107     1322925912  2018-09-22   
5          1  Санкт-Петербург       40084107     1322925912  2018-09-22   
6          1  Санкт-Петербург       40084107     1322925912  2018-09-22   
7          1  Санкт-Петербург       40084107     1236599978  2018-09-07   
8          1  Санкт-Петербург       40084107     1236599978  2018-09-07   
9          1  Санкт-Петербург       40084107     1236599978  2018-09-07   
10         1  Санкт-Петербург       40084107     1239781318  2018-09-08   
11         1  Санкт-Петербург       40084107     1239781318  2018-09-08   
12         1  Санкт-Петербург       40084107     1239781318  2018-09-08   
13         1  Санкт-Петербург       40084107     1167553543  2018-08-26   
14         1  Санкт-Петербург       40084107     1167553543  2018-08-26   
15         1  Санкт-Петербург       40084107     1167553543  2018-08-26   
16         1  Санкт-Петербург       40084107     1291100744  2018-09-17   
17         1  Санкт-Петербург       40084107     1291100744  2018-09-17   
18         1  Санкт-Петербург       40084107     1291100744  2018-09-17   
19         1  Санкт-Петербург       40084107     1237133177  2018-09-07   
20         1  Санкт-Петербург       40084107     1237133177  2018-09-07   
21         1  Санкт-Петербург       40084107     1237133177  2018-09-07   
22         1  Санкт-Петербург       40084107     1184699487  2018-08-29   
23         1  Санкт-Петербург       40084107     1184699487  2018-08-29   
24         1  Санкт-Петербург       40084107     1184699487  2018-08-29   
25         1  Санкт-Петербург       40084701     1050663231  2018-08-06   
26         1  Санкт-Петербург       40084701     1151587724  2018-08-24   
27         1  Санкт-Петербург       40084701      906282307  2018-07-12   
28         1  Санкт-Петербург       40084701     1169054092  2018-08-27   
29         1  Санкт-Петербург       40084701     1254800766  2018-09-10   
30         1  Санкт-Петербург       40084701     1317746048  2018-09-21   
31         1  Санкт-Петербург       40084701      971713131  2018-07-24   
32         1  Санкт-Петербург  4009900475495     1314186514  2018-09-21   
33         1  Санкт-Петербург  4009900475495     1042648562  2018-08-05   
34         1  Санкт-Петербург  4009900507578     1039375832  2018-08-04   
35         1  Санкт-Петербург       40099484     1192643630  2018-08-31   
36         1  Санкт-Петербург       40099484     1192643630  2018-08-31   
37         1  Санкт-Петербург       40099484     1040613322  2018-08-04   
38         1  Санкт-Петербург       40099484     1040613322  2018-08-04   
39         1  Санкт-Петербург       40099484     1358934763  2018-09-29   

                                              barname  sales_items  \
0                        ГОРОШЕКBonduelleНЕЖНЫЙ200ГЖБ          1.0   
1   ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
2   ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
3   ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
4   ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          2.0   
5   ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          2.0   
6   ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          2.0   
7   ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
8   ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
9   ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
10  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
11  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
12  ДKinderChocola

In [41]:
#Some data-explorations below
#merged_left_out[(merged_left_out['Category 2'].isnull()==True) | (merged_left_out['Category 3'].isnull()==True) | (merged_left_out['Category 4'].isnull()==True) | (merged_left_out['Category 5'].isnull()==True)]
#merged_left_out[(merged_left_out['Category 2'].isnull()==True) & (merged_left_out['Category 3'].isnull()==True) & (merged_left_out['Category 4'].isnull()==True) & (merged_left_out['Category 5'].isnull()==True)]

#merged_left_out['Category 2'].value_counts()
#merged_left_out['Category 3'].value_counts()
#merged_left_out['Category 4'].value_counts()
#merged_left_out['Category 5'].value_counts()
#merged_left_out[(merged_left_out['Category 4'] == 'Вина Франции')]
print("Count null's per variable\n", merged_left_out.isnull().sum())

Count null's per variable
 Store_id               0
Region                 0
barcode                0
TransactionID          0
Date                   0
barname                0
sales_items            0
sales_value            0
CategoryID             0
CategoryName           0
BrandID                0
BrandName              0
Category 1             0
Category 2        204836
Category 3       5042667
Category 4       7953049
Category 5       7966276
dtype: int64


In [42]:
#Fill the "unknown" for the N/A in the Category 1-5 new columns
merged_left_out.fillna("Неклассифицированные", inplace = True) 
print("Count null's per variable\n", merged_left_out.isnull().sum())

Count null's per variable
 Store_id         0
Region           0
barcode          0
TransactionID    0
Date             0
barname          0
sales_items      0
sales_value      0
CategoryID       0
CategoryName     0
BrandID          0
BrandName        0
Category 1       0
Category 2       0
Category 3       0
Category 4       0
Category 5       0
dtype: int64


In [44]:
merged_left_out['Category 1'].value_counts()

Продукты питания (folder)       3671690
Табакокурение (folder)          3314417
Неклассифицированные             365407
Unknown Category                 200608
Товары для животных (folder)     193606
                                 ...   
детские кружки                        1
Платки носовые бумажные               1
Чай фруктовый                         1
средства от плесени                   1
солёное печенье                       1
Name: Category 1, Length: 70, dtype: int64

In [48]:
#Replace Unknown Category to Неклассифицированные 
merged_left_out = merged_left_out.replace('Unknown Category', 'Неклассифицированные')
merged_left_out = merged_left_out.replace('default', 'Неклассифицированные')
merged_left_out = merged_left_out.replace('(folder)', '')  
merged_left_out['Category 1'].value_counts()

Продукты питания (folder)       3671690
Табакокурение (folder)          3314417
Неклассифицированные             566015
Товары для животных (folder)     193606
Гигиена (folder)                  95046
                                 ...   
Пюре быстрого приг.                   1
Крабовые палочки, мясо                1
средства от плесени                   1
Супы сухие для варки                  1
Галантерея (folder)                   1
Name: Category 1, Length: 69, dtype: int64

In [49]:
#!Alert!: REview this on re-run. Why "(folder)" was not stripped? Maybe best to use str.strip("(folder)"? 
merged_left_out = merged_left_out.replace('(folder)', '')

In [50]:
print("Count null's per variable\n", merged_left_out.isnull().sum())

Count null's per variable
 Store_id         0
Region           0
barcode          0
TransactionID    0
Date             0
barname          0
sales_items      0
sales_value      0
CategoryID       0
CategoryName     0
BrandID          0
BrandName        0
Category 1       0
Category 2       0
Category 3       0
Category 4       0
Category 5       0
dtype: int64


In [51]:
merged_left_out.fillna("Неклассифицированные", inplace = True)

In [52]:
print("Count null's per variable\n", merged_left_out.isnull().sum())

Count null's per variable
 Store_id         0
Region           0
barcode          0
TransactionID    0
Date             0
barname          0
sales_items      0
sales_value      0
CategoryID       0
CategoryName     0
BrandID          0
BrandName        0
Category 1       0
Category 2       0
Category 3       0
Category 4       0
Category 5       0
dtype: int64


In [53]:
merged_left_out.head(10)

Store_id           Region        barcode  TransactionID        Date  \
0         1  Санкт-Петербург  3083680538404     1138789911  2018-08-22   
1         1  Санкт-Петербург       40084107     1214906739  2018-09-03   
2         1  Санкт-Петербург       40084107     1214906739  2018-09-03   
3         1  Санкт-Петербург       40084107     1214906739  2018-09-03   
4         1  Санкт-Петербург       40084107     1322925912  2018-09-22   
5         1  Санкт-Петербург       40084107     1322925912  2018-09-22   
6         1  Санкт-Петербург       40084107     1322925912  2018-09-22   
7         1  Санкт-Петербург       40084107     1236599978  2018-09-07   
8         1  Санкт-Петербург       40084107     1236599978  2018-09-07   
9         1  Санкт-Петербург       40084107     1236599978  2018-09-07   

                                             barname  sales_items  \
0                       ГОРОШЕКBonduelleНЕЖНЫЙ200ГЖБ          1.0   
1  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
2  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
3  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
4  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          2.0   
5  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          2.0   
6  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          2.0   
7  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
8  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   
9  ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...          1.0   

   sales_value  CategoryID                                       CategoryName  \
0         86.0          65  Продукты питания (folder)/Консервы (folder)/Ко...   
1         90.0       19677  Продукты питания (folder)/Кондитерские изделия...   
2         90.0       19677  Продукты питания (folder)/Кондитерские изделия...   
3         90.0       19677  Продукты питания (folder)/Кондитерские изделия...   
4        180.0       19677  Продукты питания (folder)/Кондитерские изделия...   
5        180.0       19677  Продукты питания (folder)/Кондитерские изделия...   
6        180.0       19677  Продукты питания (folder)/Кондитерские изделия...   
7         90.0       19677  Продукты питания (folder)/Кондитерские изделия...   
8         90.0       19677  Продукты питания (folder)/Кондитерские изделия...   
9         90.0       19677  Продукты питания (folder)/Кондитерские изделия...   

   BrandID       BrandName                 Category 1  \
0    22653       Bonduelle  Продукты питания (folder)   
1    30123  Kinder Сюрприз  Продукты питания (folder)   
2    30123  Kinder Сюрприз  Продукты питания (folder)   
3    30123  Kinder Сюрприз  Продукты питания (folder)   
4    30123  Kinder Сюрприз  Продукты питания (folder)   
5    30123  Kinder Сюрприз  Продукты питания (folder)   
6    30123  Kinder Сюрприз  Продукты питания (folder)   
7    30123  Kinder Сюрприз  Продукты питания (folder)   
8    30123  Kinder Сюрприз  Продукты питания (folder)   
9    30123  Kinder Сюрприз  Продукты питания (folder)   

                      Category 2        Category 3            Category 4  \
0              Консервы (folder)  Консервы овощные  Неклассифицированные   
1  Кондитерские изделия (folder)           Шоколад  Неклассифицированные   
2  Кондитерские изделия (folder)           Шоколад  Неклассифицированные   
3  Кондитерские изделия (folder)           Шоколад  Неклассифицированные   
4  Кондитерские изделия (folder)           Шоколад  Неклассифицированные   
5  Кондитерские изделия (folder)           Шоколад  Неклассифицированные   
6  Кондитерские изделия (folder)           Шоколад  Неклассифицированные   
7  Кондитерские изделия (folder)           Шоколад  Неклассифицированные   
8  Кондитерские изделия (folder)           Шоколад  Неклассифицированные   
9  Кондитерские изделия (folder)           Шоколад  Неклассифицированные   

             Category 5  
0  Неклассифицированные  
1  Неклассифицирова

In [54]:
print ("Yahoooo!")

Yahoooo!


<hr>
<h4>Impute values for Unknown Barname and Unknown Category Name</h4>
<h5>For the unknown Barname use the lowest-level Category Name available.</h5>

In [55]:
# check out records with "Unknown" barname, as potential candidates for imputing values 
merged_left_out[(merged_left_out['barname']=='Unknown Name')].groupby(['CategoryName'],as_index= False).count()


CategoryName  Store_id  Region  \
0                                Неклассифицированные         1       1   
1                        Неклассифицированные/default        65      65   
2                      Парфюмерия (folder)/Дезодорант         1       1   
3   Продукты питания (folder)/Алкогольные напитки/...         4       4   
4   Продукты питания (folder)/Алкогольные напитки/...      1303    1303   
5              Продукты питания (folder)/Каши, хлопья        10      10   
6   Продукты питания (folder)/Кондитерские изделия...         8       8   
7   Продукты питания (folder)/Кондитерские изделия...       637     637   
8           Продукты питания (folder)/Крупа и бобовые         3       3   
9   Продукты питания (folder)/Масложировая продукц...        12      12   
10  Продукты питания (folder)/Напитки безалкогольн...       168     168   
11  Продукты питания (folder)/Напитки безалкогольн...       147     147   
12                    Продукты питания (folder)/Орехи        77      77   
13  Продукты питания (folder)/Продукты быстрого пр...       218     218   
14  Продукты питания (folder)/Рыба и морепродукты ...        80      80   
15                    Табакокурение (folder)/Сигареты       187     187   
16  Товары для животных (folder)/Корм для животных...       332     332   

    barcode  TransactionID  Date  barname  sales_items  sales_value  \
0         1              1     1        1            1            1   
1        65             65    65       65           65           65   
2         1              1     1        1            1            1   
3         4              4     4        4            4            4   
4      1303           1303  1303     1303         1303         1303   
5        10             10    10       10           10           10   
6         8              8     8        8            8            8   
7       637            637   637      637          637          637   
8         3              3     3        3            3            3   
9        12             12    12       12           12           12   
10      168            168   168      168          168          168   
11      147            147   147      147          147          147   
12       77             77    77       77           77           77   
13      218            218   218      218          218          218   
14       80             80    80       80           80           80   
15      187            187   187      187          187          187   
16      332            332   332      332          332          332   

    CategoryID  BrandID  BrandName  Category 1  Category 2  Category 3  \
0            1        1          1           1           1           1   
1           65       65         65          65          65          65   
2            1        1          1           1           1           1   
3            4        4          4           4           4           4   
4         1303     1303       1303        1303        1303        1303   
5           10       10         10          10          10          10   
6            8        8          8           8           8           8   
7          637      637        637         637         637         637   
8            3        3          3           3           3           3   
9           12       12         12          12          12          12   
10         168      168        168         168         168         168   
11         147      147        147         147         147         147   
12          77       77         77          77          77          77   
13         218      218        218         218         218         218   
14          80       80         80          80          80          80   
15         187      187        187         187         187         187   
16         332      332        332         332         332         332   

    Category 4  Category 5  
0            1           1  
1           65      

In [56]:
merged_left_out[(merged_left_out['CategoryName']=='Неклассифицированные') & (merged_left_out['barname']=='Unknown Name')]

Store_id           Region        barcode  TransactionID        Date  \
3155456       206  Курская область  4690388057312     1184560463  2018-08-29   

              barname  sales_items  sales_value  CategoryID  \
3155456  Unknown Name          1.0         19.0           0   

                 CategoryName  BrandID      BrandName            Category 1  \
3155456  Неклассифицированные        0  Unknown Brand  Неклассифицированные   

                   Category 2            Category 3            Category 4  \
3155456  Неклассифицированные  Неклассифицированные  Неклассифицированные   

                   Category 5  
3155456  Неклассифицированные

In [57]:
#Drop the one record with Unknown Category and Name
indexNames = merged_left_out[(merged_left_out['CategoryName']=='Неклассифицированные') & (merged_left_out['barname']=='Unknown Name')].index

#indexNames
# Delete these row indexes from dataFrame
merged_left_out.drop(indexNames, inplace=True)
merged_left_out[(merged_left_out['CategoryName']=='Неклассифицированные') & (merged_left_out['barname']=='Unknown Name')]

Empty DataFrame
Columns: [Store_id, Region, barcode, TransactionID, Date, barname, sales_items, sales_value, CategoryID, CategoryName, BrandID, BrandName, Category 1, Category 2, Category 3, Category 4, Category 5]
Index: []

In [58]:
#For the rest of Unknown Barnames, impute the lowest-level category available
indexNames = merged_left_out[(merged_left_out['barname']=='Unknown Name')].index
indexNames



Int64Index([ 187131,  187132,  187133,  187134,  187135,  187136,  187137,
             187138,  187139,  187140,
            ...
            7751135, 7751136, 7751137, 7751138, 7751139, 7751140, 7751141,
            7751142, 7923559, 7923560],
           dtype='int64', length=3252)

In [59]:
for index in indexNames: 
    barName = "Неклассифицированные"
    i = max_cat
    while(barName == "Неклассифицированные"):
        barName = merged_left_out["Category " +str(i)][index] 
        i = i - 1
    merged_left_out['barname'][index] = barName   

/Applications/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


In [60]:
# check the stats
merged_left_out[(merged_left_out['barname']=='Unknown Name')]

Empty DataFrame
Columns: [Store_id, Region, barcode, TransactionID, Date, barname, sales_items, sales_value, CategoryID, CategoryName, BrandID, BrandName, Category 1, Category 2, Category 3, Category 4, Category 5]
Index: []

In [61]:
print ("Yahoooo 2!")

Yahoooo 2!


<h5>For the unknown Category Name, use the Machine-Learning algorithms to predict best Category 1</h5>

<div class="alert alert-block alert-danger">
<b>Stuck at this section:</b> Unable to utilize correctly any of known to me class-prediction models. Will try couple more times, if unsucsefull - need to drop those records from further analysis (label "Unknown category"). 
 <hr>
 <strong>56 6015 records in Unknown category! Definitely point for improvement... </strong>
</div>

In [ ]:
#fit the prediction model to map the possible category, based on other features:
# - Create a copy of dataframe
# - Normalize barname 
# - Find variables with highest correlation, affecting barname
# - Split into train, test, predict dataframes
# - Use K-nearest neghbor on the variables with max correlation, to fit the model for category prediction 
# - predict category value for 14k raws with missing category 
# - update the df with predicted records 

In [62]:
# check out records with "Unknown" category as potential candidates for imputing values 
merged_left_out[(merged_left_out['CategoryName']=='Неклассифицированные')].groupby(['barname'],as_index= False).count()

barname  Store_id  Region  \
0           002КонфКарамельБаянСулуFrult7кггКостанай         3       3   
1                          002КонфЯшкиноМексикана5кг         1       1   
2      01BourjoisРумянаДвойныеBlushDuoEffetSculpteur         1       1   
3                           01ЖигулевскоеНРвкегеТула       287     287   
4                    01СамараЖивоеСветлоеНРКегРостов        15      15   
...                                              ...       ...     ...   
14002                      яйцосуперприз3в13гршантоу        54      54   
14003                                      яйцошокол         1       1   
14004                                       якобс3в1         1       1   
14005                          якобсмонарх12гкрепкий         1       1   
14006                                        янтарны         4       4   

       barcode  TransactionID  Date  sales_items  sales_value  CategoryID  \
0            3              3     3            3            3           3   
1            1              1     1            1            1           1   
2            1              1     1            1            1           1   
3          287            287   287          287          287         287   
4           15             15    15           15           15          15   
...        ...            ...   ...          ...          ...         ...   
14002       54             54    54           54           54          54   
14003        1              1     1            1            1           1   
14004        1              1     1            1            1           1   
14005        1              1     1            1            1           1   
14006        4              4     4            4            4           4   

       CategoryName  BrandID  BrandName  Category 1  Category 2  Category 3  \
0                 3        3          3           3           3           3   
1                 1        1          1           1           1           1   
2                 1        1          1           1           1           1   
3               287      287        287         287         287         287   
4                15       15         15          15          15          15   
...             ...      ...        ...         ...         ...         ...   
14002            54       54         54          54          54          54   
14003             1        1          1           1           1           1   
14004             1        1          1           1           1           1   
14005             1        1          1           1           1           1   
14006             4        4          4           4           4           4   

       Category 4  Category 5  
0               3           3  
1               1           1  
2               1           1  
3             287         287  
4              15          15  
...           ...         ...  
14002          54          54  
14003           1           1  
14004           1           1  
14005           1           1  
14006           4           4  

[14007 rows x 17 columns]

In [ ]:
#Test the linear dependancy of CategoryID with other values
cdf = merged_left_out[['Store_id','barcode','TransactionID','sales_items','sales_value','CategoryID', 'BrandID']]
cdf.head()


In [ ]:
#Check if there is correlation between barcode and category id
plt.scatter(cdf.barcode, cdf.CategoryID,  color='blue')
plt.xlabel("Barcode")
plt.ylabel("Category")
plt.show()

#!Alert! Maybe its worth to normalize data?? 

In [ ]:
#Plot the heatmap for correlation between all variables in the dataset
#plt.pcolor(cdf.corr(), cmap='RdBu')
#plt.colorbar()
#plt.show()


fig, ax = plt.subplots()
fig.subplots_adjust(bottom=0.25,left=0.25) # make room for labels

heatmap = ax.pcolor(cdf.corr(), cmap='RdBu')
cbar = plt.colorbar(heatmap)

# Set ticks in center of cells
ax.set_xticks(np.arange(cdf.corr().shape[1]) + 0.5, minor=False)
ax.set_yticks(np.arange(cdf.corr().shape[0]) + 0.5, minor=False)

# Rotate the xlabels. Set both x and y labels to headers[1:]
ax.set_xticklabels(cdf.headers[1:],rotation=90)
ax.set_yticklabels(cdf.headers[1:])

plt.show()

#Heatmap show there is low probability of linear correlation between all variables in dataset

In [ ]:
cdf.corr()['CategoryID'].sort_values() 
#insignificant correlation

In [ ]:
# try Multiple Linear regression using all variables

msk = np.random.rand(len(cdf)) < 0.8
train = cdf[msk]
test = cdf[~msk]

regr = linear_model.LinearRegression()
x = np.asanyarray(train[['Store_id','barcode','TransactionID','sales_items','sales_value', 'BrandID']])
x = preprocessing.StandardScaler().fit(x).transform(x.astype(float))


y = np.asanyarray(train[['CategoryID']])
regr.fit (x, y)

# The coefficients
print ('Coefficients: ', regr.coef_)


y_hat= regr.predict(test[['Store_id','barcode','TransactionID','sales_items','sales_value', 'BrandID']])
x = np.asanyarray(test[['Store_id','barcode','TransactionID','sales_items','sales_value', 'BrandID']])
y = np.asanyarray(test[['CategoryID']])
print("Residual sum of squares: %.2f"
      % np.mean((y_hat - y) ** 2))

# Variance score: 1 is perfect prediction
print('Variance score: %.2f' % regr.score(x, y))

#Resume: complete nonsence 

In [ ]:
#Try the categorical ML algorithms

cdf['CategoryID'].value_counts()
#Result: Total 517 category ID's 

In [ ]:
#Check how many Category 1 values
merged_left_out['Category 1'].value_counts()

#Total 70 Category 1. 
#? Maybe it'll be worth to test classification target variable Category 1, not Category ID. 
#! To check accuracy of both later 

In [ ]:
#!ALERT!: do not RUN, it hangs. Need to come up with more targeted attemt... 
#Try the K-nearest Neighbor 

#? Possibly need to add Date to CDF to improve accuracy? How date is being normalized? 
#convert date field from string to datetime
#cdf['Date'] = pd.to_datetime(cdf['InvoiceDate'])

X = cdf[['Store_id','barcode','TransactionID','sales_items','sales_value', 'BrandID']].values  #.astype(float)
y = cdf['CategoryID'].values

#Normalize
X = preprocessing.StandardScaler().fit(X).transform(X.astype(float))

#Treain/Test split 80/20%
X_train, X_test, y_train, y_test = train_test_split( X, y, test_size=0.2, random_state=4)
print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

k = 4
#Train Model and Predict  
neigh = KNeighborsClassifier(n_neighbors = k).fit(X_train,y_train)
yhat = neigh.predict(X_test)


print("Train set Accuracy: ", metrics.accuracy_score(y_train, neigh.predict(X_train)))
print("Test set Accuracy: ", metrics.accuracy_score(y_test, yhat))

In [ ]:
#group and find correlations between variables 
#Store_id, Region, barcode, TransactionID, Date, barname, sales_items, sales_value
df_group_trans = merged_left_out[['Store_id','Region','TransactionID', 'Date', 'barname', 'sales_items', 'sales_value']].groupby(['TransactionID'],as_index= False).sum()


df_group_trans.plot(kind='scatter', x='Region', y='sales_value', figsize=(10, 6), color='darkblue')

plt.title('Title')
plt.xlabel('Store')
plt.ylabel('Sales')

plt.show()

In [63]:
#Save the cleaned dataframe to CSV file
merged_left_out.to_csv('dt_challenge_cleansed.csv') 

<hr>
<h2 id="wrangle">Data Preparation</h2>
<p>Prepare all dataframes to be used in data analysis </p>

In [69]:
#Prepare cdf_apri_b and cdf_apri_cat  for the apriori analysis on the barcode and categories that are normally bought together

cdf_apri_b = merged_left_out[['TransactionID', 'Store_id', 'Region', 'Date', 'barname', 'sales_items', 'sales_value']] 
cdf_apri_cat = merged_left_out[['TransactionID', 'Store_id', 'Region', 'barcode', 'Date', 'CategoryName', 'sales_items', 'sales_value']]

#cdf_apri_b = (cdf_apri_b.groupby(['TransactionID', 'barname'])['sales_value'].sum().unstack().reset_index().fillna(0).set_index('TransactionID')) 
cdf_apri_b.head()

ValueError: Unstacked DataFrame is too big, causing int32 overflow

In [70]:
cdf_apri_b = cdf_apri_b['barname'].transpose()
cdf_apri_b.head()

0                         ГОРОШЕКBonduelleНЕЖНЫЙ200ГЖБ
1    ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...
2    ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...
3    ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...
4    ДKinderChocolateЯЙЦОKINDERСЮРПРИЗКИНДЕРМашаиМе...
Name: barname, dtype: object

In [71]:
cdf_apri_b.shape

(7966549,)

In [ ]:
#TODO: 

#df['Category 1'] = df['Category 1'].str.replace('\W', '') - need to sptrip from cat 1-5, or leave as-is. Check at DataVis state. 
#

#transpose data, so one transaction is one row
#basket = (df[df['Country'] =="France"]
#          .groupby(['InvoiceNo', 'Description'])['Quantity']
#          .sum().unstack().reset_index().fillna(0)
#          .set_index('InvoiceNo'))
#basket

<hr>
<h2 id="model">Data Modelling</h2>
<p>Model the data, using Python/ML algorithms.</p>

<div class="alert alert-block alert-info">
<b>Q1:</b> Найти самые часто встречающиеся комбинации <code> продуктов </code> и <code> категорий </code> в чеках (что люди обычно покупают вместе).
</div>


In [ ]:
#Use Apriori algorithm to answer the 1st question: https://www.edureka.co/blog/apriori-algorithm/#apriori-algorithm 

<hr>
<h2 id="vis">Results and Findings</h2>
<p>Presenting insights from Data Analysis and recommendations</p>

<div class="alert alert-block alert-info">
<b>Q2:</b> Выяснить, насколько сильно различаются потребительские привычки по разным регионам и с чем это может быть связано.
</div>


<div class="alert alert-block alert-info">
<b>Q3:</b> Описать профили магазинов, в которых средние продажи той или иной категории (в штуках или деньгах), значимо выше или ниже относительно среднего уровня.
</div>


In [ ]:
#Vis 3: Bar plot/hist
#y = amount in rubbles
#x = geo regions
#legend = brands 
#bars by category .. possibly impossible... 

#df_s.plot(kind='bar', figsize=(15, 6), color=['#5cb85c', '#5bc0de', '#d9534f']) 
#plt.title('Percentage of the Respondents Interest in Data Science Areas') 
# annotate value labels to each response 
#for i, value in enumerate(df_s): 
    #for j, value2 in enumerate(df_s[value]): 
        #label = format(int(value2), ',') 
        #plt.annotate(value2, xy=((i*0.18 + j - 0.25), value2 + 1), color='black') 
        
#plt.show()

In [ ]:
#Vis 4: By highest and lowest performing regions, get the wordcloud of their categories. 
#Replicate by Highest and lowest performing Store's and Brands.  

#Vis 5: For each highest and lowest region above, plot histograms: 
# X = store_id 
# y = amount in rub 
# Legend of hist stacks = top 3 category

#Vis 6: Replicate vis 5 with x = top and bottom 10 brand 
# X = brand
# y = amount in rub 
# Legend of hist stacks = top 3 category

In [ ]:
#Vis 7: 
# Bubble Plot 
# x = count unique transaction id's for each region, normalize 
# y = sum all item counts for each region (normalize?)
# z = sum all amounts for each region (normalize?)
# See overall for all categories, zoom in by-categories top and bottom 3 categorie

<hr>
<h2 id="app">Appendixes</h2>
<p>Installing packages and updates (anaconda)</p>

In [1]:
!conda install -c anaconda xlrd --yes
!conda update -n base -c defaults conda --yes

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.7.12
  latest version: 4.8.0rc0

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /Applications/anaconda3

  added / updated specs:
    - xlrd


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2019.8.28  |                0         133 KB  anaconda
    certifi-2019.9.11          |           py37_0         154 KB  anaconda
    conda-4.7.12               |           py37_0         3.0 MB  anaconda
    openssl-1.1.1d             |       h1de35cc_2         3.4 MB  anaconda
    ------------------------------------------------------------
                                           Total:         6.7 MB

The following packages will be SUPERSEDED by a higher-priority channel:

  ca-certificates                